src/memory-management/comparison.md

```diff
@@ -24,12 +24,12 @@ Here is a rough comparison of the memory management techniques.
   * Double-frees.
   * Memory leaks.
 * Automatic like Java:
+  * Runtime overhead.
   * Garbage collection pauses.
   * Destructor delays.
 * Scope-based like C++:
-  * Complex, opt-in by programmer (on C++).
-  * Circular references can lead to memory leaks
```

Comment:
This one is very important I think

Reply:
I removed it because it applies to all the techniques.

In CPython, reference counting is used for garbage collection. In addition, there is a [garbage collector](https://docs.python.org/3/library/gc.html) which runs once in a while to detect and clean up reference cycles. As described on that page, you can disable this garbage collector if you know you never create cycles. It used to be the case that cycles between objects with a destructor (a `__del__` method) would be uncollectable, but this is apparently no longer true after [PEP 442](https://peps.python.org/pep-0442/). If I understand correctly, it is now only C-level types which can become uncollectable.

In any case, the argument about reference cycles leaking memory also applies to manual memory management and it even applies to Rust. So I don't think we need to highlight it here.

Reply:
It is inherent to reference counting: you cannot build reference cycles in (safe) rust at all, unless if you use `Rc` (which is the same thing as a `shared_ptr`/ARC again), and a GC will also clean up cycles.

Reply:
@domenukk Reading your comment I realize that the course equates automatic memory management to garbage collection (in summary.md, automatic memory management is called garbage collection).

In my mental model, automatic memory management could be thought of as an interface for the programmer that allows them to not care about the memory lifecycles. While GC vs RC is an implementation detail.

Scope based forces the programmer to care about the lifecycles through move and reference semantics while abstracting away the actual allocations and destructions. So the interface is different for the programmer.

Maybe the class doesn't need to mention Garbage Collection or reference counting at all and just focus on automatic vs manual vs RAII?

Reply:
Yes maybe GC and RefCounting overcomplicate things then, sorry for the confusion.

As non-c++ person, thinking of RAII as a memory management model was non-obvious.
To make my confusion worse, the concept of RAII can be implemented using RefCounting, and wiki isn't talking about memory at all...
https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization#Reference_counting